File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    first = None
    second = None
    flag = False
    prev = None
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None:
            return root
        
        self.inorder(root)
        temp = self.first.val
        self.first.val = self.second.val
        self.second.val = temp
    
    def inorder(self, root):
        if root is None:
            return 
        
        self.inorder(root.left)
        #st.pop()
        if(self.prev != None and self.prev.val >= root.val):
            if(self.flag == False):
                self.first = self.prev
                self.second = root
                self.flag = True
            else:
                self.second = root
        
        self.prev = root
        self.inorder(root.right)

File: 0099-recover-binary-search-tree/0099-recover-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode first;
    TreeNode second;
    TreeNode prev;
    boolean flag;
    public void recoverTree(TreeNode root) {
        if(root == null) return;
        inorder(root);

        int temp = first.val;
        first.val = second.val;
        second.val = temp;
    }

    private void inorder(TreeNode root){
        //base
        if(root == null){
            return;
        }
        //logic
        inorder(root.left);
        if(prev != null && prev.val >= root.val){
            if(!flag){
                first = prev;
                second = root;
                flag = true;
            }
            else{
                second = root;
            }
        }

        prev = root;
        inorder(root.right);
    }
}

//DFS
//Time Complexity: O(n)
//Space Complexity: O(n)
